crates/query-engine/translation/src/translation/helpers.rs

```diff
@@ -90,16 +91,71 @@ pub enum CollectionInfo<'env> {
 }
 
 #[derive(Debug)]
-/// Metadata information about a specific collection or composite type.
+/// Metadata information about a composite type. This includes both tables (and views) and
+/// dedicated composite types.
 pub enum CompositeTypeInfo<'env> {
-    CollectionInfo(CollectionInfo<'env>),
-    CompositeTypeInfo {
-        name: String,
-        info: metadata::CompositeType,
+    Table {
+        name: &'env str,
+        info: &'env metadata::TableInfo,
+    },
+    CompositeType {
+        name: &'env str,
+        info: &'env metadata::CompositeType,
+    },
+}
+
+#[derive(Debug)]
+/// Metadata information about any object that can have fields
+pub enum FieldsInfo<'env> {
+    Table {
+        name: &'env str,
+        info: &'env metadata::TableInfo,
+    },
+    NativeQuery {
+        name: &'env str,
+        info: &'env metadata::NativeQueryInfo,
+    },
+    CompositeType {
+        name: &'env str,
+        info: &'env metadata::CompositeType,
     },
 }
 
+impl<'a> From<&'a CompositeTypeInfo<'a>> for FieldsInfo<'a> {
+    fn from(value: &'a CompositeTypeInfo<'a>) -> Self {
+        match value {
+            CompositeTypeInfo::Table { name, info } => FieldsInfo::Table { name, info },
+            CompositeTypeInfo::CompositeType { name, info } => {
+                FieldsInfo::CompositeType { name, info }
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a CollectionInfo<'a>> for FieldsInfo<'a> {
+    fn from(value: &'a CollectionInfo<'a>) -> Self {
+        match value {
+            CollectionInfo::Table { name, info } => FieldsInfo::Table { name, info },
+            CollectionInfo::NativeQuery { name, info } => FieldsInfo::NativeQuery { name, info },
+        }
+    }
+}
+
 impl<'request> Env<'request> {
+    pub fn with_empty<F, R>(f: F) -> R
```

Comment:
This has to have some docs.

Reply:
Ah, agreed. It's a bit non-trivial what and why we do this.